### PR TITLE
Add OpenTherm component (part 3: rest of the sensors)

### PR DESCRIPTION
**Pull request by olegtarasov**
# What does this implement/fix?

This is the third (and final) part of OpenTherm component. It contains the code to support all the sensors, inputs and outputs. Sensors configured in YAML are transformed to a set of C++ macros, which in turn generate the final code that handles sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Part 1: #6645 
- Part 2: #7529 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#3800

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# An extremely minimal configuration which only enables you to set the boiler's
# water temperature setpoint as a number.

esphome:
  name: thermostat-number-minimal

esp8266:
  board: d1_mini

opentherm:
  in_pin: IN_PIN
  out_pin: OUT_PIN

number:
  - platform: opentherm
    t_set:
      name: "Boiler Control setpoint"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by olegtarasov** [comment]
Hi @jesserockz! The final part of OpenTherm integration is ready for review. Thanks!

**Review by jesserockz — COMMENTED** [review]
How does the hub know if a switch (or other control type entity) has been changed by the user? I do not see the `add_on_state_callback`s being added.

I find it very difficult to follow and understand this component due to the heavy use of macros.

**Comment by olegtarasov** [comment]
@jesserockz The hub doesn't need to know the exact moment when control entity changes because all sensors, switches and outputs are sent to OpenTherm device constantly in a loop. It's required by OpenTherm standard to communicate with device every 1 second even when there is no change in controls. So the approach when we send all the messages configured in yaml in a loop adds to component's robustness (a message can fail due to protocol error, and we don't need to use complicated retry logic — the same message will be sent again eventually).

> I find it very difficult to follow and understand this component due to the heavy use of macros.

I agree, but I see no other way to add class members dynamically, and adding > 100 class members for all the available sensors at once seems very wasteful. What's more, most of them likely won't be supported by end user's boiler.

File: tests/components/opentherm/common.yaml [hunk 1]
```diff
@@ -74,3 +105,86 @@ sensor:
...
+      kp: 0.7
+      ki: 0.003
+  - platform: pid
+    name: "Central heating (Circuit 2)"
+    heat_output: t_set_ch2
+    default_target_temperature: 18
+    sensor: ch2_room_temperature
+    control_parameters:
+      kp: 0.4
+      ki: 0.004
```
Comment (jesserockz) [inline tests/components/opentherm/common.yaml]: Adding this here wont achieve anything and if the `pid` component has a breaking change, these files wont be run. They should be contained to only itself and dependencies
```suggestion
```

File: tests/components/opentherm/common.yaml [hunk 2]
```diff
@@ -74,3 +105,86 @@ sensor:
...
+    entity_id: sensor.temperature
+    filters:
+      # Push room temperature every second to update PID parameters
+      - heartbeat: 1s
+  - platform: homeassistant
+    id: ch2_room_temperature
+    entity_id: sensor.other_temperature
+    filters:
+      # Push room temperature every second to update PID parameters
+      - heartbeat: 1s
```
Comment (jesserockz) [inline tests/components/opentherm/common.yaml]: Same with these
```suggestion
```

File: tests/components/opentherm/common.yaml [hunk 3]
```diff
@@ -12,10 +12,41 @@ opentherm:
   cooling_enable: false
   otc_active: false
   ch2_active: true
+  t_room: ch_room_temperature
```
Comment (jesserockz) [inline tests/components/opentherm/common.yaml]: Just use one of the opentherm sensors here as this is a simple compile test and doesnt actually run

File: esphome/components/opentherm/switch/switch.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  } else {
+    ESP_LOGD("opentherm.switch", "Restored state for OpenTherm switch '%s': %d", this->get_name().c_str(),
+             restored.value());
+    state = restored.value();
+  }
+  this->write_state(state);
+}
+
+void OpenthermSwitch::dump_config() {
+  esphome::switch_::log_switch("opentherm.switch", "", "OpenTherm Switch", this);
```
Comment (jesserockz) [inline esphome/components/opentherm/switch/switch.cpp]:
```suggestion
  LOG_SWITCH(TAG, "", "OpenTherm Switch", this);
```

File: esphome/components/opentherm/switch/switch.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#include "switch.h"
+
+namespace esphome {
+namespace opentherm {
+
```
Comment (jesserockz) [inline esphome/components/opentherm/switch/switch.cpp:5]:
```suggestion

static const char *const TAG = "opentherm.switch";

```

File: esphome/components/opentherm/switch/switch.cpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+void OpenthermSwitch::write_state(bool state) { this->publish_state(state); }
+
+void OpenthermSwitch::setup() {
+  auto restored = this->get_initial_state_with_restore_mode();
+  bool state = false;
+  if (!restored.has_value()) {
+    ESP_LOGD("opentherm.switch", "Couldn't restore state for OpenTherm switch '%s'", this->get_name().c_str());
```
Comment (jesserockz) [inline esphome/components/opentherm/switch/switch.cpp]: Please update to add TAG to all the cpp files and use that for logging
```suggestion
    ESP_LOGD(TAG, "Couldn't restore state for OpenTherm switch '%s'", this->get_name().c_str());
```

File: esphome/components/opentherm/switch/switch.cpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  } else {
+    ESP_LOGD("opentherm.switch", "Restored state for OpenTherm switch '%s': %d", this->get_name().c_str(),
```
Comment (jesserockz) [inline esphome/components/opentherm/switch/switch.cpp]:
```suggestion
    ESP_LOGD(TAG, "Restored state for OpenTherm switch '%s': %d", this->get_name().c_str(),
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**